File: leetcode-java/src/main/java/chapter05/pointer/IntersectArray.java
package chapter05.pointer;

import java.util.*;

/**
 * <p>1. 两个数组交集</p>
 * <p>2. 两个数组交集 II</h3>
 */
public class IntersectArray {

    /**
     * <p>两个数组的交集</p>
     * <p>1. 哈希表</p>
     * <p>2. 双指针</p>
     */
    public static class ArrayIntersectI {

        /**
         * 哈希表
         * <p>只需要将第一个数组放入哈希表, 然后遍历第二个数组, 只要在哈希表中存在就放入交集, 然后移除元素</p>
         */
        public static int[] intersectionHash(int[] firstNum, int[] secondNum) {
            int current = 0;
            int[] intersection = new int[firstNum.length + secondNum.length];
            Set<Integer> set = new HashSet<>();
            for (int index = 0; index < firstNum.length; index++) {
                set.add(firstNum[index]);
            }
            for (int index = 0; index < secondNum.length; index++) {
                if (set.contains(secondNum[index])) {
                    intersection[current++] = secondNum[index];
                    set.remove(secondNum[index]);
                }
            }
            return Arrays.copyOf(intersection, current);
        }

        /**
         * 双指针
         * <p>如果内存有限, 那么只能采用双指针的方式, 采用外部排序 + 双指针</p>
         */
        public static int[] intersection(int[] firstNum, int[] secondNum) {
            // 1. 排序
            Arrays.sort(firstNum);
            Arrays.sort(secondNum);
            // 2. 双指针遍历
            int index = 0;
            int first = 0, second = 0;
            int[] intersection = new int[firstNum.length + secondNum.length];
            while (first < firstNum.length && second < secondNum.length) {
                if (firstNum[first] < secondNum[second]) {
                    first++;
                } else if (firstNum[first] > secondNum[second]) {
                    second++;
                } else {
                    if (first == 0 || firstNum[first] != firstNum[first - 1]) {
                        intersection[index++] = firstNum[first];
                    }
                    first++;
                    second++;
                }
            }
            return Arrays.copyOf(intersection, index);
        }

    }

    /**
     * <p>两个数组的交集 II</p>
     */
    public static class ArrayIntersectII {

        /**
         * 哈希表
         * <p>只需要将第一个数组放入哈希表, 然后遍历第二个数组, 只要在哈希表中存在就放入交集, 然后移除元素</p>
         */
        public static int[] intersectionHash(int[] firstNum, int[] secondNum) {
            int current = 0;
            int[] intersection = new int[firstNum.length + secondNum.length];
            Map<Integer, Integer> map = new HashMap<>();
            for (int index = 0; index < firstNum.length; index++) {
                map.put(firstNum[index], map.getOrDefault(firstNum[index] ,0 ) + 1);
            }
            for (int index = 0; index < secondNum.length; index++) {
                int count = map.getOrDefault(secondNum[index], 0);
                if (count > 0) {
                    intersection[current++] = secondNum[index];
                    count--;
                    map.put(secondNum[index], count);
                }
            }
            return Arrays.copyOf(intersection, current);
        }

        /**
         * 双指针
         * <p>如果内存有限, 那么只能采用双指针的方式, 采用外部排序 + 双指针</p>
         */
        public static int[] intersection(int[] firstNum, int[] secondNum) {
            // 1. 排序
            Arrays.sort(firstNum);
            Arrays.sort(secondNum);
            // 2. 双指针
            int index = 0;
            int first = 0, second = 0;
            int[] intersection = new int[firstNum.length + secondNum.length];
            while (first < firstNum.length && second < secondNum.length) {
                if (firstNum[first] < secondNum[second]) {
                    first++;
                } else if (firstNum[first] > secondNum[second]) {
                    second++;
                } else {
                    intersection[index++] = firstNum[first];
                    first++;
                    second++;
                }
            }
            return Arrays.copyOf(intersection, index);
        }

    }

}

File: leetcode-go/chapter05/pointer/intersect_array.go
package pointer

import "sort"

// 两个数组的交集
type ArrayIntersectI struct {}

// 哈希表
func (array *ArrayIntersectI) intersectionHash(nums1, nums2 []int) []int {
	intersection := make([]int, 0)
	set := make(map[int]struct{})
	for index := 0; index < len(nums1); index++ {
		set[nums1[index]] = struct{}{}
	}
	for index := 0; index < len(nums2); index++ {
		if _, ok := set[nums2[index]]; ok {
			intersection = append(intersection, nums2[index])
			delete(set, nums2[index])
		}
	}
	return intersection
}

// 双指针
func (array *ArrayIntersectI) intersection(nums1, nums2 []int) []int {
	sort.Ints(nums1)
	sort.Ints(nums2)
	first, second := 0, 0
	intersection := make([]int, 0)
	for first < len(nums1) && second < len(nums2) {
		if nums1[first] < nums2[second] {
			first++
		} else if nums1[first] > nums2[second] {
			second++
		} else {
			if first == 0 || nums1[first] != nums1[first - 1] {
				intersection = append(intersection, nums1[first])
			}
			first++
			second++
		}
	}
	return intersection
}

// 两个数组的交集 II
type ArrayIntersectII struct {}

// 哈希表
func (array *ArrayIntersectII) intersectionHash(nums1, nums2 []int) []int {
	intersection := make([]int, 0)
	count := make(map[int]int)
	for index := 0; index < len(nums1); index++ {
		count[nums1[index]]++
	}
	for index := 0; index < len(nums2); index++ {
		if count[nums2[index]] > 0 {
			intersection = append(intersection, nums2[index])
			count[nums2[index]]--
		}
	}
	return intersection
}

// 双指针
func (array *ArrayIntersectII) intersection(nums1, nums2 []int) []int {
	sort.Ints(nums1)
	sort.Ints(nums2)
	first, second := 0, 0
	intersection := make([]int, 0)
	for first < len(nums1) && second < len(nums2) {
		if nums1[first] < nums2[second] {
			first++
		} else if nums1[first] > nums2[second] {
			second++
		} else {
			intersection = append(intersection, nums1[first])
			first++
			second++
		}
	}
	return intersection
}